### PR TITLE
Change Set/Map IterAll to use LoadLeafSequences

### PR DESCRIPTION
**Pull request by kalman**

### CODE REVIEW

File: go/types/list.go [hunk 1]
```diff
@@ -170,33 +173,39 @@ func (l List) IterAll(f listIterAllFunc) {
...
+	d.PanicIfFalse(startIdx < l)
 
-	if startIdx == endIdx {
-		return
+	endIdx := startIdx + blockLength
+	d.PanicIfFalse(endIdx <= l)
+
+	valuesPerIdx := uint64(1)
+	if col.Kind() == MapKind {
+		valuesPerIdx = 2
```
Comment (arv) [inline go/types/list.go]: you could make this part of the collection interface.

Comment (kalman) [reply go/types/list.go]: eh doesn't seem worthwhile updating all those call sites. I made it a function though.

File: go/types/leaf_sequence.go [hunk 2]
```diff
@@ -72,10 +72,11 @@ func (seq leafSequence) valuesSlice(from, to uint64) []Value {
 	if len := seq.Len(); to > len {
 		to = len
 	}
+
 	dec := seq.decoderSkipToIndex(int(from))
-	vs := make([]Value, to-from)
-	for i := from; i < to; i++ {
-		vs[i-from] = dec.readValue()
+	vs := make([]Value, (to-from)*getValuesPerIdx(seq))
```
Comment (kalman) [inline go/types/leaf_sequence.go]: Updating this made me realise that `values()` had a bug for Maps, we just never called it for Maps.

File: go/types/list.go [hunk 3]
```diff
@@ -170,33 +173,35 @@ func (l List) IterAll(f listIterAllFunc) {
...
-		copy(out, values)
-		out = out[len(values):]
-
-		endIdx = endIdx - uint64(len(values)) - startIdx
+		numValues += len(values)
+		endIdx = endIdx - uint64(len(values))/valuesPerIdx - startIdx
 		startIdx = 0
-		numBytes += uint64(len(ls.buff))
+
+		// TODO: this is wrong, it should only include the bytes of |values| only,
```
Comment (rafael-atticlabs) [inline go/types/list.go]: probably not wrong enough to worry about.

File: go/types/list.go [hunk 4]
```diff
@@ -170,33 +173,35 @@ func (l List) IterAll(f listIterAllFunc) {
 	}
 }
 
-func (l List) copyReadAhead(out []Value, startIdx uint64) (numBytes uint64) {
-	llen := l.Len()
-	d.PanicIfFalse(startIdx < llen)
+func copyReadAhead(col Collection, blockLength uint64, startIdx uint64) (out []Value, numBytes uint64) {
```
Comment (rafael-atticlabs) [inline go/types/list.go]: let's make this `iterRange(f func(V Value), start, end)` or `getRange(start, end)`. you choose.

File: go/types/map.go [hunk 5]
```diff
@@ -222,12 +221,16 @@ func (m Map) IteratorFrom(key Value) MapIterator {
 type mapIterAllCallback func(key, value Value)
 
 func (m Map) IterAll(cb mapIterAllCallback) {
-	cur := newCursorAt(m.orderedSequence, emptyKey, false, false, true)
-	cur.iter(func(v interface{}) bool {
-		entry := v.(mapEntry)
-		cb(entry.key, entry.value)
-		return false
+	var k Value
+	iterAll(m, func(v Value, idx uint64) {
```
Comment (rafael-atticlabs) [inline go/types/map.go]: nice. this makes sense.

File: go/types/list.go (hunk 3 above)
Comment (kalman) [reply go/types/list.go]: ok

Comment (kalman) [reply go/types/list.go]: (kept a little note there, for future reference)

**Pull request merged**